### PR TITLE
Add reshape_as_blocks function

### PR DESCRIPTION
**Pull request by larrybradley**
This is often a handy function to have.  Using this function, I refactored `block_reduce` to remove astropy's dependency on `scikit-image`.  An additional bonus is `block_reduce` is faster now.

I split the `block_*` functions from `nddata/utils.py` into a new module called `nddata/blocks.py`.
The first few commits are necessary git gymnastics to preserve the history of both files.

### CODE REVIEW

**Comment by pllim** [comment]
Technically, this didn't make the 4.1 feature freeze for new PRs last Friday but I'll defer to @bsipocz .

**Comment by larrybradley** [comment]
@bsipocz noted that the `matplotlib` docs dependency had been being pulled in by `scikit-image`.  I explicitly added `matplotlib` as a docs dependency.

**Comment by bsipocz** [comment]
@pllim - we haven't branched, so whatever is merged this week will be certainly in (as I'm not willing to mess with the history for the sake of forcing a double deadline)

File: CHANGES.rst [hunk 1]
```diff
@@ -79,6 +79,8 @@ astropy.nddata
   when specifying the bit flags to be used or ignored when converting a bit
   field to a boolean. [#10095]
 
+- Added ``reshape_as_blocks`` function. [#10214]
```
Comment (astrofrog) [inline CHANGES.rst:82]: Can you give more context here as to what it does?

Comment (larrybradley) [reply CHANGES.rst]: I added some more context.  It does exactly what it says on the tin.  :smile:

**Comment by larrybradley** [comment]
Because I split the `block_*` functions from `utils.py` into a new `blocks.py` module, importing these functions from `nddata.utils` would no longer work:  e.g., `from astropy.nddata.utils import block_reduce` would fail.  I just added a commit that allows the old imports from `nddata.utils` to still work, but issues a deprecation warning about the new module.  The module-level `__class__` reassignment can be replaced with PEP562 when the minimum required python version is 3.7.

**Comment by bsipocz** [comment]
I think v4.2 will have minimun python 3.7, so a cleanup PR can be opened once we're done with 4.1

**Comment by bsipocz** [comment]
(maybe not right away, as keeping the 3.6 tests around for a bit will significantly help with the 4.1.1 bugfix release)

File: astropy/nddata/tests/test_blocks.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import numpy as np
+import pytest
+
+from astropy.nddata import reshape_as_blocks, block_reduce, block_replicate
+
+
+class TestReshapeAsBlocks:
```
Comment (pllim) [inline astropy/nddata/tests/test_blocks.py:9]: Add a test to check that `reshaped` is really a view and not a copy?

File: astropy/nddata/blocks.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    block_size_int = block_size.astype(int)
+    if np.any(block_size_int != block_size):  # e.g., 2.0 is OK, 2.1 is not
+        raise ValueError('block_size elements must be integers')
+
+    return data, block_size_int
+
+
+def reshape_as_blocks(data, block_size):
+    """
+    Reshape a data array into blocks.
```
Comment (pllim) [inline astropy/nddata/blocks.py:36]: This needs a "version added" directive?

Comment (larrybradley) [reply astropy/nddata/blocks.py:36]: I can add that, but I didn't think astropy used it very much.

Comment (pllim) [reply astropy/nddata/blocks.py:36]: Its usage is not consistent throughout the package, but doesn't hurt?

File: astropy/nddata/blocks.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    block_size = np.atleast_1d(block_size)
+    if data.ndim > 1 and len(block_size) == 1:
+        block_size = np.repeat(block_size, data.ndim)
+
+    if len(block_size) != data.ndim:
+        raise ValueError('block_size must be a scalar or have the same '
+                         'length as the number of data dimensions')
+
+    if np.any(block_size <= 0):
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply astropy/nddata/blocks.py]: Yes, done.

File: astropy/nddata/blocks.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    --------
+    >>> import numpy as np
+    >>> from astropy.nddata import block_replicate
+    >>> data = np.array([[0., 1.], [2., 3.]])
+    >>> block_replicate(data, 2)  # doctest: +FLOAT_CMP
+    array([[0.  , 0.  , 0.25, 0.25],
+           [0.  , 0.  , 0.25, 0.25],
+           [0.5 , 0.5 , 0.75, 0.75],
+           [0.5 , 0.5 , 0.75, 0.75]])
+
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply astropy/nddata/blocks.py:193]: I actually wanted the blank line to separate the two examples.

File: astropy/nddata/tests/test_blocks.py (hunk 2 above)
Comment (larrybradley) [reply astropy/nddata/tests/test_blocks.py:9]: Done.

File: astropy/nddata/blocks.py (hunk 3 above)
Comment (larrybradley) [reply astropy/nddata/blocks.py:36]: Done.

File: astropy/nddata/blocks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pllim) [reply astropy/nddata/blocks.py:193]: Hmm, as long as it renders fine, I guess it's ok.

**Comment by pllim** [comment]
I'll wait for the docs to render and have a look before approving, otherwise LGTM. Thanks!

File: astropy/nddata/blocks.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    -------
+    output : array_like
+        The resampled data.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from astropy.nddata import block_reduce
+    >>> data = np.arange(16).reshape(4, 4)
+    >>> block_reduce(data, 2)    # doctest: +SKIP
```
Comment (pllim) [inline astropy/nddata/blocks.py:133]: Wait... why is it still skipped but my comment marked as resolved? Hope you can clarify. Thanks.

Comment (larrybradley) [reply astropy/nddata/blocks.py:133]: Hmm, I accepted your suggestion above.

Comment (larrybradley) [reply astropy/nddata/blocks.py]: I just fixed that (again).

File: astropy/nddata/blocks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (larrybradley) [reply astropy/nddata/blocks.py:193]: It does.

File: astropy/nddata/blocks.py (hunk 3 above)
Comment (bsipocz) [reply astropy/nddata/blocks.py:36]: >  I didn't think astropy used it very much

not yet, but hopefully we change that 😅

**Comment by pllim** [comment]
Please don't merge yet. I am gonna have one last look, since Larry did ask me to review. 😉

**Comment by larrybradley** [comment]
Here are some timings comparing astropy 4.0 (old) and this PR (new).  For 1000x1000 arrays, the new version is ~2x faster.  For 5000x5000 arrays, the new version is ~4-10x faster (resulting in ~0.12 s speedup).

```python
>>> import numpy as np
>>> from astropy.nddata import block_reduce
>>> np.random.seed(123)
>>> data = np.random.random((1000, 1000))
>>> %timeit block_reduce(data, (10, 10))
old: 2.30 ms    new: 1.48 ms    ratio: 1.55
>>> %timeit block_reduce(data, (50, 50))
old: 1.39 ms    new: 669 µs    ratio: 2.08
>>> %timeit block_reduce(data, (100, 100))
old: 1.32 ms    new: 580 µs    ratio: 2.28

>>> data = np.random.random((5000, 5000))
>>> %timeit block_reduce(data, (10, 10))
old: 160 ms    new: 37.6 ms    ratio: 4.26
>>> %timeit block_reduce(data, (50, 50))
old: 139 ms    new: 14.6 ms    ratio: 9.52
>>> %timeit block_reduce(data, (100, 100))
old: 138 ms    new: 13.8 ms    ratio: 10.0
```

**Comment by pllim** [comment]
Nice! Should we add this to https://github.com/astropy/astropy-benchmarks ?

**Pull request merged**